R/verb-fill.R

```diff
@@ -0,0 +1,169 @@
+#' @examples
+#' squirrels <- tibble::tribble(
+#'   ~group,    ~name,     ~role,     ~n_squirrels, ~ n_squirrels2,
+#'   1,      "Sam",    "Observer",   NA,                 1,
+#'   1,     "Mara", "Scorekeeper",    8,                NA,
+#'   1,    "Jesse",    "Observer",   NA,                NA,
+#'   1,      "Tom",    "Observer",   NA,                 4,
+#'   2,     "Mike",    "Observer",   NA,                NA,
+#'   2,  "Rachael",    "Observer",   NA,                 6,
+#'   2,  "Sydekea", "Scorekeeper",   14,                NA,
+#'   2, "Gabriela",    "Observer",   NA,                NA,
+#'   3,  "Derrick",    "Observer",   NA,                NA,
+#'   3,     "Kara", "Scorekeeper",    9,                 10,
+#'   3,    "Emily",    "Observer",   NA,                NA,
+#'   3, "Danielle",    "Observer",   NA,                NA
+#' )
+#' squirrels$id <- 1:12
+#'
+#' dbplyr_fill(
+#'   tbl_memdb(squirrels),
+#'   n_squirrels,
+#'   n_squirrels2,
+#'   order_by = c(id)
+#' )
+dbplyr_fill <- function(.data, ..., order_by, .direction = c("down", "up")) {
+  # Sources
+  # * https://www.xspdf.com/resolution/51601218.html
+  # * https://dzone.com/articles/how-to-fill-sparse-data-with-the-previous-non-empt
+
+  sim_data <- simulate_vars(.data)
+  cols_to_fill <- syms(names(tidyselect::eval_select(expr(c(...)), sim_data)))
+  order_by_cols <- enquo(order_by)
+  .direction <- arg_match0(.direction, c("down", "up"))
+
+  if (.direction == "up") {
+    order_by_cols <- quo(-!!order_by_cols)
+  }
+
+  dbplyr_fill_int(
+    .con = remote_con(.data),
+    .data = .data,
+    cols_to_fill = cols_to_fill,
+    order_by_cols = order_by_cols,
+    .direction = .direction
+  )
+}
+
+dbplyr_fill_int <- function(.con, .data, cols_to_fill, order_by_cols, .direction) {
+  UseMethod("dbplyr_fill_int")
+}
+
+#' @export
+dbplyr_fill_int.SQLiteConnection <- function(.con,
+                                             .data,
+                                             cols_to_fill,
+                                             order_by_cols,
+                                             .direction) {
+  partition_sql <- purrr::map(
+    cols_to_fill,
+    ~ translate_sql(
+      cumsum(ifelse(is.na(!!.x), 0L, 1L)),
+      vars_order = translate_sql(!!order_by_cols, con = .con),
+      vars_group = op_grps(.data),
+    )
+  ) %>%
+    set_names(paste0("..dbplyr_partion_", seq_along(cols_to_fill)))
+
+  dp <- .data %>%
+    mutate(!!!partition_sql)
+
+  fill_sql <- purrr::map2(
+    cols_to_fill, names(partition_sql),
+    ~ translate_sql(
+      max(!!.x, na.rm = TRUE),
+      con = .con,
+      vars_order = translate_sql(!!order_by_cols, con = .con),
+      vars_group = c(op_grps(.data), .y),
+    )
+  ) %>%
+    set_names(purrr::map_chr(cols_to_fill, as_name))
+
+  dp %>%
+    transmute(
+      !!!syms(colnames(.data)),
+      !!!fill_sql
+    ) %>%
+    select(!!!colnames(.data))
+}
+
+#' @export
+dbplyr_fill_int.PostgreSQL <- dbplyr_fill_int.SQLiteConnection
```

Comment:
I have mixed feelings about putting all these methods in this file. On one hand, it makes it easier to understand and review, but on the other hand, it makes it harder to see how to add a new backend, and doesn't follow the existing conventions. Maybe I should review like this, and then spread out after the fact?

Reply:
yes, totally makes sense. The easier understanding was the reason I started it this way and then I forgot to split it into the different backend files...